receiver/telemetrygeneratorreceiver/otlp_generator.go

```diff
@@ -0,0 +1,214 @@
+func findLastTraceEndTime(traces ptrace.Traces) time.Time {
+	var maxTime time.Time
+	for i := 0; i < traces.ResourceSpans().Len(); i++ {
+		resourceSpans := traces.ResourceSpans().At(i)
+		for k := 0; k < resourceSpans.ScopeSpans().Len(); k++ {
+			scopeSpans := resourceSpans.ScopeSpans().At(k)
+			for j := 0; j < scopeSpans.Spans().Len(); j++ {
+				span := scopeSpans.Spans().At(j)
+				if span.EndTimestamp().AsTime().After(maxTime) {
+					maxTime = span.EndTimestamp().AsTime()
+					continue
+				}
+				if maxTime.IsZero() {
+					maxTime = span.EndTimestamp().AsTime()
+				}
+			}
+		}
+	}
+	return maxTime
+}
+
+// adjustTraceTimes changes the start and end times of all spans to be relative to the current time, placing
+// the span that ends at maxTime at the current time.
+func (g *otlpGenerator) adjustTraceTimes(maxTime time.Time) {
+	now := getCurrentTime()
+	for i := 0; i < g.traces.ResourceSpans().Len(); i++ {
+		resourceSpans := g.traces.ResourceSpans().At(i)
+		for k := 0; k < resourceSpans.ScopeSpans().Len(); k++ {
+			scopeSpans := resourceSpans.ScopeSpans().At(k)
+			for j := 0; j < scopeSpans.Spans().Len(); j++ {
+				span := scopeSpans.Spans().At(j)
+				// delta is the differnce between this spans end time and the max end time
+				delta := maxTime.Sub(span.EndTimestamp().AsTime())
+				// spanDuration is the length of the span
+				spanDuration := span.EndTimestamp().AsTime().Sub(span.StartTimestamp().AsTime())
+				endTime := now
+				// move each span's end time by delta
+				span.SetEndTimestamp(pcommon.NewTimestampFromTime(endTime.Add(delta)))
+				// set the start time to be the end time minus the original span duration
+				span.SetStartTimestamp(pcommon.NewTimestampFromTime(endTime.Add(-spanDuration)))
+			}
+		}
+	}
+	// save the current time we used as a baseline to adjust the spans
+	g.tracesStart = now
+}
+
+func (g *otlpGenerator) generateLogs() plog.Logs {
+	for i := 0; i < g.logs.ResourceLogs().Len(); i++ {
+		resourceLogs := g.logs.ResourceLogs().At(i)
+		for k := 0; k < resourceLogs.ScopeLogs().Len(); k++ {
+			scopeLogs := resourceLogs.ScopeLogs().At(k)
+			for j := 0; j < scopeLogs.LogRecords().Len(); j++ {
+				log := scopeLogs.LogRecords().At(j)
+				log.SetTimestamp(pcommon.NewTimestampFromTime(getCurrentTime()))
+			}
+		}
+	}
+	return g.logs
+}
+
+func (g *otlpGenerator) generateMetrics() pmetric.Metrics {
+	for i := 0; i < g.metrics.ResourceMetrics().Len(); i++ {
+		resourceMetrics := g.metrics.ResourceMetrics().At(i)
+		for k := 0; k < resourceMetrics.ScopeMetrics().Len(); k++ {
+			scopeMetrics := resourceMetrics.ScopeMetrics().At(k)
+			for j := 0; j < scopeMetrics.Metrics().Len(); j++ {
+				metric := scopeMetrics.Metrics().At(j)
+				switch metric.Type() {
+				case pmetric.MetricTypeSum:
+					dps := metric.Sum().DataPoints()
+					for l := 0; l < dps.Len(); l++ {
+						dps.At(l).SetTimestamp(pcommon.NewTimestampFromTime(getCurrentTime()))
```

Comment:
Likely all of these metrics should have the same timestamp. We can probably call `getCurrentTime` once at the top and use it everywhere.